#' matrixQTL
#'
#' This is the wrapper function for the eQTL/aQTL analyses using the MatrixEQTL
#' package, which was authored and is maintained by Andrey A Shabalin
#' and is under an LGPL-3 license. For MatrixEQTL troubleshooting and bug
#' reports, visit https://github.com/andreyshabalin/MatrixEQTL.
#'
#' @param snp_dose string indicating the path and file name for the SNP dosage
#' file.
#' @param snp_map string indicating the path and file name for the SNP map file.
#' @param gene_data string indicating the path and file name for the gene
#' exression/activities file.
#' @param gene_map string indicating the path and file name for the gene map
#' file.
#' @param covars string indicating the path and file name for the covariables
#' file. If you do not wish to include covariables in the QTL analysis (not
#' typically recommended), do not assign anything to this argument. Defaults to
#' NULL.
#' @param prefix string indicating a prefix to be added to all output file
#' names.
#' @param cisP numeric indicating the P-value threshold for cis-QTLs to be
#' written to the output file. Defaults to 1 (all cis-QTL results are output).
#' @param transP numeric indicating the P-value threshold for trans-QTLs to be
#' written to the output file. Defaults to 0.05 (only nominally significant
#' QTLs are output).
#'
#' @return Writes four output files:
#' \enumerate{
#'   \item  A log file recording the standard outputs generated by the function.
#'   \item  A pdf file for the QQ-plot of all tested cis and trans-QTLs.
#'   \item  A text file with the cis-QTL summary statistics that includes only
#'   those that passed the cis P threshold set by the cisP argument.
#'   \item  A text file with the trans-QTL summary statistics that includes
#'   only those that passed the trans P threshold set by the transP argument.
#' }
#' @export
#'
#' @examples
#' \dontrun{
#' matrixQTL(
#' "GEUVADIS_filtered_samples_WHR_significant_bi-allelic_SNPs.dosage",
#' "GEUVADIS_WHR_significant_bi-allelic_SNPs_locations_in_GRCh37.map",
#' "Toy_data_regulators_activities.txt",
#' "GEUVADIS_expressed_genes_locations_in_GRCh37.map",
#' "GEUVADIS_filtered_samples_select_covars.txt",
#' prefix="Toy_data")
#' }
matrixQTL <- function(snp_dose,snp_map,gene_data,gene_map,covars=NULL,prefix,cisP=1,transP=0.05){

  # Send standard output to a log file.
  sink(paste(prefix,"_MatrixQTL_std_out.log",sep = ""),split = T)

  ## Settings

  # Linear model to use, modelANOVA, modelLINEAR, or modelLINEAR_CROSS
  useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS

  # Genotype file name
  SNP_file_name = snp_dose;
  snps_location_file_name = snp_map;

  # Gene expression file name
  expression_file_name = gene_data;
  gene_location_file_name = gene_map;

  # Covariates file name
  # Set to character() for no covariates
  if(is.null(covars)){
    covariates_file_name = character();
  } else{
    covariates_file_name = covars;
  }

  # Output file name
  output_file_name_cis = tempfile();
  output_file_name_tra = tempfile();

  # Only associations significant at this level will be saved
  pvOutputThreshold_cis = as.numeric(cisP);
  pvOutputThreshold_tra = as.numeric(transP);

  # Error covariance matrix
  # Set to numeric() for identity.
  errorCovariance = numeric();
  # errorCovariance = read.table("Sample_Data/errorCovariance.txt");

  # Distance for local gene-SNP pairs
  cisDist = 1e6;

  ## Load genotype data

  snps = SlicedData$new();
  snps$fileDelimiter = "\t";      # the TAB character
  snps$fileOmitCharacters = "NA"; # denote missing values;
  snps$fileSkipRows = 1;          # one row of column labels
  snps$fileSkipColumns = 1;       # one column of row labels
  snps$fileSliceSize = 2000;      # read file in slices of 2,000 rows
  snps$LoadFile(SNP_file_name);

  ## Load gene expression data

  gene = SlicedData$new();
  gene$fileDelimiter = "\t";      # the TAB character
  gene$fileOmitCharacters = "NA"; # denote missing values;
  gene$fileSkipRows = 1;          # one row of column labels
  gene$fileSkipColumns = 1;       # one column of row labels
  gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows
  gene$LoadFile(expression_file_name);


  ## Normal quantile transformation of gene expression data

  for( sl in 1:length(gene) ) {
    mat = gene[[sl]];
    mat = t(apply(mat, 1, rank, ties.method = "average"));
    mat = stats::qnorm(mat / (ncol(gene)+1));
    gene[[sl]] = mat;
  }
  rm(sl, mat);


  ## Load covariates

  cvrt = SlicedData$new();
  cvrt$fileDelimiter = "\t";      # the TAB character
  cvrt$fileOmitCharacters = "NA"; # denote missing values;
  cvrt$fileSkipRows = 1;          # one row of column labels
  cvrt$fileSkipColumns = 1;       # one column of row labels
  if(length(covariates_file_name)>0) {
    cvrt$LoadFile(covariates_file_name);
  }

  ## Run the analysis
  snpspos = utils::read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);
  genepos = utils::read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE);

  me = Matrix_eQTL_main(
    snps = snps,
    gene = gene,
    cvrt = cvrt,
    output_file_name     = output_file_name_tra,
    pvOutputThreshold     = pvOutputThreshold_tra,
    useModel = useModel,
    errorCovariance = errorCovariance,
    verbose = TRUE,
    output_file_name.cis = output_file_name_cis,
    pvOutputThreshold.cis = pvOutputThreshold_cis,
    snpspos = snpspos,
    genepos = genepos,
    cisDist = cisDist,
    pvalue.hist = "qqplot",
    min.pv.by.genesnp = FALSE,
    noFDRsaveMemory = FALSE);

  unlink(output_file_name_tra);
  unlink(output_file_name_cis);

  ## Results:

  cat('Analysis done in: ', me$time.in.sec, ' seconds', '\n');
  cis_eqtls<-me$cis$eqtls
  cat('Detected local QTLs:', dim(cis_eqtls)[1],'\n');
  trans_eqtls<-me$trans$eqtls
  cat('Detected distant QTLs:', dim(trans_eqtls)[1],'\n');

  ## Plot the Q-Q plot of local and distant p-values
  grDevices::pdf(paste(prefix,"_QQ_plot.pdf",sep = ""))
  plot(me)
  grDevices::dev.off()

  utils::write.table(cis_eqtls,paste(prefix,"_cis_results_",cisP,"_P_threshold.txt",sep = ""),sep="\t",quote = FALSE,row.names=FALSE)
  utils::write.table(trans_eqtls,paste(prefix,"_trans_results_",transP,"_P_threshold.txt",sep = ""),sep="\t",quote = FALSE,row.names=FALSE)

  closeAllConnections()

}
